README
Term-ReadLine-Bender 0.07
=========================

ABOUT

  This package provides a set of modules that form an interactive input buffer
  written in plain perl with minimal dependencies. It features almost all
  key-bindings described in the posix spec for the sh(1) utility with some
  extensions like multiline editing; this includes a vi-command mode with a
  save-buffer (for copy-pasting) and an undo-stack.

  It is compatible with the Term::ReadLine interface, so it
  can be used with perl programs using this interface.

DEPENDENCIES

  The only real dependencies for this package are Term::ReadKey and AutoLoader
  which are available from cpan.

  One function in the vi command mode needs File::Temp, but as long as you don't
  use this function you don't need the module.

  You might also want to consider to install Env::PS1 if you want a fancy prompt.

INSTALL

  Try something like:

    $ perl Build.PL
    $ ./Build test
    $ ./Build install

  See the Module::Build documentation for advanced build options.


Changes
Revision history for Term-ReadLine-Bender

0.07 Fri Jun 1 2012
   Forked from Term::ReadLine::Zoid to Term::ReadLine::Bender
   Maintainer switched to Eric Waters <ewaters@gmail.com>
   - Attribs->{lines_preprocess_function} is called before print
     This is most useful for coloring the output
   - Allow Attribs->{RPS1} to be a subroutine for dynamic right handed
     PS1 (i.e., timestamp)
   - Allow ISearch to be exited by Ctrl-C even if locked

Revision history for Term-ReadLine-Zoid - another ReadLine package

0.07 Mon Nov 22 2004
	Maintenance release, just one bugfix

     Fri Oct 22
	- Fixed a bug with the completion of quoted strings

0.06 Wed Sep 15 2004
        - Fixed a bug in the clear_screen() function
	- Touched documentation for FileBrowse
	- Changed return_empty_string logic
	- Added T:RL:Z::Emacs with 'emacs' and 'emacs_multiline' modes
	- Added ^O (operate_and_get_next)

0.05 Mon Sep  6 2004
        Release to accompany the 0.90 release of Zoidberg to Adam.pm
        - Made 'beat' also an attribute

     Sun Sep  5
        - Made it possible for "quote" meta field to be a CODE reference so
        the application can handle escaping in a custom way

     Mon Aug 30
        - Some bugfixes in FileBrowse
	- Modified behaviour of ^D for multiline buffers
	- Added basic support for buffers that have more lines then the screen

     Mon Aug  9
        - Reverse direction of arrays in SetHistory to be compatible with
	T::RL::Gnu's behaviour
	- Added the FileBrowse (fbrowse) mode

     Wed Aug  4
        - Added support to set attributes through the PERL_RL variable
	- Added the 'default_mode' attribute
	- Refactored the way keymaps and modes are handled
	- Renamed most key routines with functional names
	- Incorporated T:RL:Z:MultiLine into T:RL:Zoid, not all modii need to
	have a .pm anymore, there still is a 'multiline' keymap of course

     Tue Aug  3
        - Fixed behaviour of 'e' and 'E' in ViCommand
	- ISearch does now repeated search at ^R
	- Changed behaviour on terminal resize a bit

     Tue Jul 20
        - Added automatic logic for the 'maxcomplete' option

     Mon Jul 12
        - Fixed 'quoted' vs 'quote' logic for advanced completion

0.04 Sun Jul  4 2004
        - Made cursor position take control chars into account
	- Reversed the output of GetHistory after testing with cpan(1)

     Thu Jul  1
        - Fixed a bug, escaped returns are now really removed
	- Fixed a bug, expansion works again in a non-zoid application,
	like cpan(1)
	- Added pager support for long lists of completions

0.03 Tue Jun 29 2004
        - Fixed a bug in the expansion code regarding escaping
	- Fixed 'bindchr' and 'bindkey' vi commands
	- Added code for '=', '*', and '\' keybindings in the vi mode

0.02 Tue May 25 2004

     Mon May 24
        - Added support for more advanced completion functions

     Sun May 23
        - Added "RPS1" setting after reading something about zsh on /.
	- Made it possible for the prompts to be a scalar reference

     Fri May 21
        - Fixed an AUTOLOADER inheritance problem leading to bugs in zoid

     Sun May 16
        - Added some simple emacs bindings like ^F, ^B etc.

     Tue Mar 30
        - Fixed typo in the synopsis, patch by Mike Castle

0.01 Sun Mar 28 2004
	Initial feature set is complete enough to release

     Thu Jan 15 2004
	Initialised the module


lib/Term/ReadLine/Bender/Emacs.pm
package Term::ReadLine::Bender::Emacs;

use strict;
use base 'Term::ReadLine::Bender';

our $VERSION = 0.01;

=head1 NAME

Term::ReadLine::Bender::Emacs - a readline emacs mode

=head1 SYNOPSIS

This class is used as a mode under L<Term::ReadLine::Bender>,
see there for usage details.

=head1 DESCRIPTION

This mode provides some emacs key-bindings, taking the L<bash>(1)
implementation as a reference.

This module also provides a 'emac_multiline' key map.

=head1 KEY MAPPING

These bindings are additional to those in L<Term::ReadLine::Bender> which
already contains some emacs key bindings.

=over 4

=cut

our %_keymap = (
	ctrl_Q	=> 'quoted_insert',
	escape	=> 'prefix_meta',
	meta_f	=> 'forward_word',
	meta_b	=> 'backward_word',
	ctrl_X  => 'prefix_key',
	ctrl_X_ctrl_V => 'switch_mode_command',
	_isa	=> 'insert',
);

our %_m_keymap;

sub keymap {
	return \%_keymap unless $_[1] =~ /multiline/;
	unless (%_m_keymap) {
		%_m_keymap = %_keymap;
		$_m_keymap{_isa} = 'multiline';
	}
	return \%_m_keymap;
}

=item escape, ^[  (I<prefix_meta>)

=cut

sub prefix_meta { $_[0]->prefix_key('meta') }

sub prefix_key {
	my ($self, $pre) = @_;
	my ($key, $cnt);
	until ($key) {
		my $k = $self->read_key();
		if ($k =~ /^[\-\d]+$/) { $cnt .= $k }
		else { $key = $self->key_name( $k ) }
	}
	$self->do_key($pre.'_'.$key, $cnt);
}

=item meta-f  (I<forward_word>)

=item meta-b  (I<backward_word>)

=cut

sub forward_word { # simple version of vi_E
	my ($self, undef, $cnt) = @_;
	$cnt ||= 1;
	my $l = $$self{lines}[ $$self{pos}[1] ];
	for (1 .. $cnt) {
		if ($l =~ /^.{$$self{pos}[0]}(\w?.*?\w+)/) { $$self{pos}[0] += length($1) }
		else {
			$self->end_of_line();
			last;
		}
	}
	return 1;
}

sub backward_word { # simple version of vi_B
	my ($self, undef, $cnt) = @_;
	$cnt ||= 1;
	my $l = $$self{lines}[ $$self{pos}[1] ];
	for (1 .. $cnt) {
		$l = substr($l, 0, $$self{pos}[0]);
		if ($l =~ /(\w+[^\w]*)$/) { $$self{pos}[0] -= length $1 }
		else {
			$self->beginning_of_line;
			last;
		}
	}
	return 1;
}

1;

__END__

=item ^X^V  (I<switch_mode_command>)

Enter (vi) command mode. Taken from L<zsh>(1).

=item ^V, ^Q  (I<quoted_insert>)

Insert next key literally, ignoring any key-bindings.

WARNING: control or escape chars in the editline can cause unexpected results

=back

=head1 TODO

Get count args right (see bash reference)

A lot more bindings

A emacs multiline mode

=head1 AUTHOR

Jaap Karssenberg (Pardus) E<lt>pardus@cpan.orgE<gt>

Copyright (c) 2004 Jaap G Karssenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Term::ReadLine::Bender>

=cut



lib/Term/ReadLine/Bender/ISearch.pm
package Term::ReadLine::Bender::ISearch;

use strict;
use base 'Term::ReadLine::Bender';
no warnings; # undef == '' down here

our $VERSION = 0.06;

our %_keymap = (
	backspace  => 'backward_delete_char',
	ctrl_R     => 'isearch_again',
	_on_switch => 'is_switch',
	_default   => 'self_insert'
);

sub keymap { return \%_keymap }

sub is_switch {
	my $self = shift;
	$$self{is_lock} = undef;
	$$self{is_hist_p} = -1;
	$$self{is_save} = [[''], [0,0], undef];
}

sub is_switch_back {
	my ($self, $key) = @_;
	$$self{_hist_save} = $self->save();
	@$self{qw/lines pos hist_p/} = @{$$self{is_save}};
	$self->switch_mode();
	$self->do_key($key);
}

sub draw { # rendering this inc mode is kinda consuming
	my ($self, @args) = @_;
	my $save = $self->save();
	my $string = join "\n", @{$$self{lines}};
	$$self{prompt} = "i-search qr($string): ";
	goto DRAW unless length $string;

	my ($result, $match, $hist_p) = (undef, '', $$self{is_hist_p});
	$$self{last_search} = ['b', $string];
	my $reg = eval { qr/^(.*?$string)/ };
	goto DRAW if $@;

	while ($hist_p < $#{$$self{history}}) {
		$hist_p++;
		next unless $$self{history}[$hist_p] =~ $reg;
		($result, $match) = ($$self{history}[$hist_p], $1);
		last;
	}

	if (defined $result) {
		push @{$$self{last_search}}, $hist_p;
		$$self{is_lock} = undef;
		$$self{lines} = [ split /\n/, $result ];
		my @match = split /\n/, $match;
		$$self{pos} = [length($match[-1]), $#match];
	}
	else { $$self{is_lock} = 1 }

	DRAW: Term::ReadLine::Bender::draw($self, @args);
	$$self{is_save} = [ $$self{lines}, $$self{pos}, $hist_p];
	$self->restore($save);
}

sub self_insert {
	my ($self, $chr) = @_;
	if ($self->{is_lock} && $chr !~ /^ctrl_/) {
		$self->bell;
	}
	elsif ($self->key_binding($chr, $self->{config}{default_mode})) {
		goto \&is_switch_back;
	}
	else {
		goto \&Term::ReadLine::Bender::self_insert;
	}
}

sub isearch_again {
	my $self = shift;
	if ($self->{last_search}) {
		$self->{is_hist_p} = $self->{last_search}[-1];
	}
}

1;

__END__

=head1 NAME

Term::ReadLine::Bender::ISearch - a readline incremental search mode

=head1 SYNOPSIS

This class is used as a mode under L<Term::ReadLine::Bender>,
see there for usage details.

=head1 DESCRIPTION

This mode is intended as a work alike for the incremental search
found in the gnu readline library.

In this mode the string you enter is regarded as a B<perl regex> which is used
to do an incremental histroy search.

Pressing '^R' repeatingly will give alternative results.

Special keys like movements or the C<return> drop you out of this mode
and set the edit line to the last search result.

=head1 AUTHOR

Jaap Karssenberg || Pardus [Larus] E<lt>pardus@cpan.orgE<gt>

Copyright (c) 2004 Jaap G Karssenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

=cut



lib/Term/ReadLine/Bender/Base.pm
package Term::ReadLine::Bender::Base;

use strict;
no warnings;
use Term::ReadKey qw/ReadMode ReadKey GetTerminalSize/;
#use encoding 'utf8';
no warnings; # undef == '' down here

our $VERSION = '0.06';

$| = 1;

our @_key_buffer;

our %chr_map = ( # partial sequences
	"\e"    => '',
	"\e["   => '',
	"\eO"   => '',
	"\e[["  => '',
	( map {("\e[$_" => '')} (1 .. 24) ),

	"\e[2"	 => '',	"\e[5"	 => '',
	"\eO2"	 => '',	"\eO5"	 => '',
	"\e[1;2" => '',	"\e[1;5" => '',
);

our %chr_names = ( # named keys
	"\e"   => 'escape',
	"\cH"  => 'backspace',
	"\cI"  => 'tab',
	"\cJ"  => 'return',    # line feed
	"\cM"  => 'return',    # carriage return
	"\c?"  => 'backspace', # traditionally known as DEL

	"\e[A" => 'up',			"\eOA" => 'up',
	"\e[B" => 'down', 		"\eOB" => 'down',
	"\e[C" => 'right',		"\eOC" => 'right',
	"\e[D" => 'left', 		"\eOD" => 'left',
	"\e[F" => 'end',		"\eOF" => 'end',
	"\e[H" => 'home',		"\eOH" => 'home',

	"\e[1~"  => 'home',
	"\e[2~"  => 'insert',
	"\e[3~"  => 'delete',
	"\e[4~"  => 'end',
	"\e[5~"  => 'page_up',
	"\e[6~"  => 'page_down',
	"\e[11~" => 'f1',		"\eOP" => 'f1',		"\e[[A" => 'f1',
	"\e[12~" => 'f2',		"\eOQ" => 'f2',		"\e[[B" => 'f2',
	"\e[13~" => 'f3',		"\eOR" => 'f3',		"\e[[C" => 'f3',
	"\e[14~" => 'f4',		"\eOS" => 'f4',		"\e[[D" => 'f4',
	"\e[15~" => 'f5',					"\e[[E" => 'f5',
	"\e[17~" => 'f6',
	"\e[18~" => 'f7',
	"\e[19~" => 'f8',
	"\e[20~" => 'f9',
	"\e[21~" => 'f10',
	"\e[23~" => 'f11',
	"\e[24~" => 'f12',

	"\e[2A" => 'shift_up',		"\eO2A" => 'shift_up',		"\e[1;2A" => 'shift_up',
	"\e[2B" => 'shift_down',	"\eO2B" => 'shift_down',	"\e[1;2B" => 'shift_down',
	"\e[2C" => 'shift_right',	"\eO2C" => 'shift_right',	"\e[1;2C" => 'shift_right',
	"\e[2D" => 'shift_left',	"\eO2D" => 'shift_left',	"\e[1;2D" => 'shift_left',
	"\e[2F" => 'shift_end',		"\eO2F" => 'shift_end',		"\e[1;2F" => 'shift_end',
	"\e[2H" => 'shift_home',	"\eO2H" => 'shift_home',	"\e[1;2H" => 'shift_home',

	"\e[5A" => 'ctrl_up',		"\eO5A" => 'ctrl_up',		"\e[1;5A" => 'ctrl_up',
	"\e[5B" => 'ctrl_down',		"\eO5B" => 'ctrl_down',		"\e[1;5B" => 'ctrl_down',
	"\e[5C" => 'ctrl_right',	"\eO5C" => 'ctrl_right',	"\e[1;5C" => 'ctrl_right',
	"\e[5D" => 'ctrl_left',		"\eO5D" => 'ctrl_left',		"\e[1;5D" => 'ctrl_left',
	"\e[5F" => 'ctrl_end',		"\eO5F" => 'ctrl_end',		"\e[1;5F" => 'ctrl_end',
	"\e[5H" => 'ctrl_home',		"\eO5H" => 'ctrl_home',		"\e[1;5H" => 'ctrl_home',
);

#	'[6A' => 'ctrl_shift_up',	'O6A' => 'ctrl_shift_up',	'[1;6A' => 'ctrl_shift_up',
#	'[6B' => 'ctrl_shift_down',	'O6B' => 'ctrl_shift_down',	'[1;6B' => 'ctrl_shift_down',
#	'[6C' => 'ctrl_shift_right',	'O6C' => 'ctrl_shift_right',	'[1;6C' => 'ctrl_shift_right',
#	'[6D' => 'ctrl_shift_left',	'O6D' => 'ctrl_shift_left',	'[1;6D' => 'ctrl_shift_left',
#	'[6F' => 'ctrl_shift_end',	'O6F' => 'ctrl_shift_end',	'[1;6F' => 'ctrl_shift_end',
#	'[6H' => 'ctrl_shift_home',	'O6H' => 'ctrl_shift_home',	'[1;6H' => 'ctrl_shift_home',

#	'[7A' => 'ctrl_alt_up',		'O7A' => 'ctrl_alt_up',		'[1;7A' => 'ctrl_alt_up',
#	'[7B' => 'ctrl_alt_down',	'O7B' => 'ctrl_alt_down',	'[1;7B' => 'ctrl_alt_down',
#	'[7C' => 'ctrl_alt_right',	'O7C' => 'ctrl_alt_right',	'[1;7C' => 'ctrl_alt_right',
#	'[7D' => 'ctrl_alt_left',	'O7D' => 'ctrl_alt_left',	'[1;7D' => 'ctrl_alt_left',
#	'[7F' => 'ctrl_alt_end',	'O7F' => 'ctrl_alt_end',	'[1;7F' => 'ctrl_alt_end',
#	'[7H' => 'ctrl_alt_home',	'O7H' => 'ctrl_alt_home',	'[1;7H' => 'ctrl_alt_home',

# ############## #
# base functions #
# ############## #

sub bell {
	#print STDERR 'bell called by: ',join(', ', caller)."\n";
	exists( $_[0]{config}{bell} )
		? $_[0]{config}{bell}->()
		: print { $_[0]{OUT} } "\cG" ; # ^G == \007 == BELL
	return 0;
}

sub loop {
	my $self = shift;
	$$self{lines} = [''] unless @{$$self{lines}};
	$$self{term_size} = [ (GetTerminalSize($$self{IN}))[0,1] ];
	@ENV{'COLUMNS', 'LINES'} = @{$$self{term_size}} if $$self{config}{autoenv};
	$self->draw();
	$$self{_loop} = 1;
	while ($$self{_loop}) {
		$self->do_key();
		while (@_key_buffer) { $self->do_key() }
		$self->draw();
	}
	$self->cursor_at(@{$$self{_buffer_end}});
}

sub beat { $_[0]{config}{beat}->() if exists $_[0]{config}{beat} }

sub read_key { die "deprecated warning" if $_[1];
	my $self = shift;
	return shift @_key_buffer if scalar @_key_buffer;

	my $chr;
	ReadMode('raw', $$self{IN});
	{
		local $SIG{WINCH} = sub { $$self{_SIGWINCH} = 1 };

		while ( not defined ($chr = ReadKey(1, $$self{IN})) ) { $self->beat() }

		my $n_chr;
		if (
			exists $chr_map{$chr} and
			( $$self{config}{low_latency} or ($n_chr = ReadKey(0.05, $$self{IN})) )
		) {
			$chr .= $n_chr;
			while (exists $chr_map{$chr}) {
				while ( not defined ($n_chr = ReadKey(1, $$self{IN})) ) { $self->beat() }
				$chr .= $n_chr;
			}
			unless (exists $chr_names{$chr}) {
				$chr =~ s/^(.)(.*)/$1/s;
				push @_key_buffer, split '', $2;
			}
		}
	}
	ReadMode('normal', $$self{IN});

	return $chr;
}

sub do_key {
	my ($self, $key) = (shift, shift);
	$key = $self->read_key() unless length $key;

	# $self->key_name()
	if (exists $chr_names{$key}) { $key = $chr_names{$key} }
	elsif (length $key < 2) {
		my $ord = ord $key;
		$key =	  ($ord < 32)   ? 'ctrl_'  . (chr $ord + 64)
			: ($ord == 127) ? 'ctrl_?' : $key ;
	}

	# $self->key_binding
	my $map = $$self{keymaps}{$$self{mode}};
	my $sub;
	DO_KEY:
	if (exists $$map{$key}) { $sub = $$map{$key} }
	elsif (exists $$map{_isa}) {
		$map = $$self{keymaps}{ $$map{_isa} }
			|| return warn "$$map{_isa}: no such keymap\n\n";
		goto DO_KEY;
	}
	elsif (exists $$map{_default}) { $sub = $$map{_default} }
	else { $sub = 'bell' }

	#print STDERR "# key: $key sub: $sub\n";
	my $re = ref($sub) ? $sub->($self, $key, @_) : $self->$sub($key, @_) ;
	$$self{last_key} = $key;
	return $re;
}

sub print {
	# The idea is to let the terminal render the line wrap
	# but calculate what it will do in order to get the cursor position right.
	my ($self, $lines, $pos) = @_;
#	use Data::Dumper;
#	print STDERR Dumper $lines, $pos;
	if ($$self{_SIGWINCH}) { # GetTerminalSize is kind of heavy
		$$self{term_size} = [ (GetTerminalSize($$self{IN}))[0,1] ];
		@ENV{'COLUMNS', 'LINES'} = @{$$self{term_size}} if $$self{config}{autoenv};
		$$self{_SIGWINCH} = 0;
	}

	my ($width, $higth) = @{$$self{term_size}};

	# calculate how line wrap will work out
	my @nlines = map { int(print_length($_) / $width) } @$lines;
	$$pos[1] += $nlines[$_] for 0 .. $$pos[1] - 1;
	$$pos[1] += int($$pos[0] / $width);
	$$pos[0] %= $width;
#	print STDERR Dumper \@nlines, $pos;

	# get the lines at the right position
	my $buffer = -1; # always 1 lines minimum
	$buffer += 1 + $_ for @nlines;
	my $null = 1;
	if ($buffer > $higth) { # big buffer or small screen :$
		# FIXME does not yet reckon with line wrap
		# FIXME some +1 or -1 offsets not right
		my $offset = $$pos[1] - $$self{scroll_pos};
		if ($offset < 0) { $$self{scroll_pos} = $$pos[1] }
		elsif ($offset > $higth) { $$self{scroll_pos} += $offset - $higth }
		@$lines = splice @$lines, $$self{scroll_pos}, $higth;
		$$pos[1] -= $$self{scroll_pos};
		$$self{_buffer_end} = [$width, $higth];
		$$self{_buffer} = $higth;
	}
	else { # normal readline buffer
		if ($buffer > $$self{_buffer}) { # clear screen area
			$self->cursor_at(@{$$self{term_size}});
			print { $$self{OUT} } "\n" x ($buffer - $$self{_buffer});
			$$self{_buffer} = $buffer;
		}
		$null = $$self{term_size}[1] - $$self{_buffer};
		$$self{_buffer_end} = [print_length($$lines[-1]), $null + $buffer]; # save real cursor
	}
	$self->cursor_at(1, $null);
	print { $$self{OUT} } $$lines[$_], "\e[K\n" for 0 .. $#$lines - 1;
	print { $$self{OUT} } $$lines[-1], "\e[J";

	$self->cursor_at($$pos[0]+1, $$pos[1]+$null); # set cursor
}

# ######### #
# utilities #
# ######### #

sub TermSize { (GetTerminalSize($_[0]{IN}))[0,1] }

sub key_name {
	if (exists $chr_names{$_[1]}) { return $chr_names{$_[1]} }
	elsif (length $_[1] < 2) {
		my $ord = ord $_[1];
		return	  ($ord < 32)   ? 'ctrl_'  . (chr $ord + 64)
			: ($ord == 127) ? 'ctrl_?' : $_[1] ;
	}
	else { return $_[1] }
}

sub key_binding {
	my ($self, $key, $mode) = @_;
	$mode ||= $$self{mode};

	my $map = $$self{keymaps}{$mode};
	FIND_KEY:
	if (exists $$map{$key}) { return $$map{$key} }
	elsif (exists $$map{_isa}) {
		$map = $$self{keymaps}{ $$map{_isa} }
			|| return warn "$$map{_isa}: no such keymap\n\n";
		goto FIND_KEY;
	}
	else { return undef }
}

sub press {
	my $self = shift;
	push @_key_buffer, (@_>1) ? (@_) : (split '', $_[0]);
	while (scalar @_key_buffer) { $self->do_key() }
}

sub unread_key {
	my $self = shift;
	unshift @_key_buffer, (@_>1) ? (@_) : (split '', $_[0]);
}

sub pos2off {
	my ($self, $pos) = @_;
	$pos ||= $$self{pos};
	my $off = $$pos[0];
	$off += 1 + length $$self{lines}[$_] for 0 .. $$pos[1] - 1;
	return $off;
}

sub output {
	my ($self, @items) = @_;

	$self->cursor_at(@{$$self{_buffer_end}});
	print { $$self{OUT} } "\n";

	my ($max, $cnt) = ($$self{config}{maxcomplete}, scalar @items);
	$self->_ask($cnt) or return if $max and $max =~ /^\d+$/ and $cnt > $max;

	@items = ($cnt > 1) ? ($self->col_format(@items)) : (split /\n/, $items[0]);

	$$self{_buffer} = (@items < $$self{_buffer}) ? ($$self{_buffer} - @items) : 0;
	if (@items > $$self{term_size}[1]) {
		$self->_ask($cnt) or return if $max and $max eq 'pager';
		my $pager = $ENV{PAGER} || 'more';
		eval {
			local $SIG{PIPE} = 'IGNORE';
			open PAGER, "| $pager" || die;
			print PAGER join("\n", @items), "\n";
			close PAGER;
		} ;
	}
	else { print { $$self{OUT} } join("\n", @items), "\n" }
}

sub _ask {
	my ($self, $cnt) = @_;
	print { $$self{OUT} } "Display all $cnt possibilities? [yN]";
	my $answ = $self->read_key();
	print { $$self{OUT} } "\n";
	return( ($answ =~ /y/i) ? 1 : 0 );
}

sub col_format { # takes minimum number of rows, but fills cols first
	my ($self, @items) = @_;

	my $len = 0;
	$_ > $len and $len = $_ for map {length $_} @items;
	$len += 2; # spacing
	my ($width) = $self->TermSize();
	return @items if $width < (2 * $len); # rows == items
	return join '  ', @items if $width > (@items * $len); # 1 row

	my $cols = int($width / $len ) - 1; # 0 based
	my $rows = int(@items / ($cols+1)); # 0 based ceil
	$rows -= 1 unless @items % ($cols+1); # tune ceil
	my @rows;
	for my $r (0 .. $rows) {
		my @row = map { $items[ ($_ * ($rows+1)) + $r] } 0 .. $cols;
		push @rows, join '', map { $_ .= ' 'x($len - length $_) } @row;
	}
	#print STDERR scalar(@items)." items, $len long, $width width, $cols+1 cols, $rows+1 rows\n";
	return @rows;
}

# ################# #
# Key binding stuff #
# ################# #

sub bindchr {
	my ($self, $chr, $key) = @_;
	if ($chr =~ /^\^(.)$/) { $chr = eval qq/"\\c$1"/ }
	$chr_names{$chr} = $key;
	chop $chr;
	while (length $chr) {
		$chr_map{$chr} = '';
		chop $chr;
	}
}

sub recalc_chr_map {
	my $self = shift;
	%chr_map = ();
	while (my ($k,$v) = each %chr_names) {
		$self->bindchr($k, $v);
	}
}

# ########## #
# ANSI stuff #
# ########## #

sub cursor_at { print { $_[0]{OUT} } "\e[$_[2];$_[1]H" } # ($x, $y) 1-based !

sub new_line {
	my $self = shift;
	return unless -t $$self{OUT} and -t $$self{IN};

	ReadMode 'raw';
	my $r;
	print { $$self{OUT} } "\e[6n";
	$r = ReadKey( -1, $$self{IN}) || return print { $$self{OUT} } "\n";
	while ($r =~ /^(\e|\e\[\d*|\e\[\d+;\d*)$/) { $r .= ReadKey -1, $$self{IN} }
	# in this case timed read doesn't work :(
	ReadMode 'normal';

	if ($r =~ /^\e\[\d+;(\d+)\D$/) {
		print { $$self{OUT} } "\n" if $1 > 1;
	}
	else {
		$self->unread_key($r);
		print { $$self{OUT} } "\n";
	}
}

sub clear_screen { print { $_[0]{OUT} } "\e[2J" }

sub print_length {
	my $string = pop;
	$string =~ s{\e\[[\d;]*\w}{}g; # strip ansi codes
	return length $string;
}

## Sequences from the "How to change the title of an xterm" howto
##  <http://www.tldp.org/HOWTO/Xterm-Title.html>
sub title {
	my ($self, $title) = @_;
	return unless $ENV{TERM};
	my $string =
		($ENV{TERM} =~ /^((ai)?xterm.*|dtterm|screen)$/) ? "\e]0;$title\cG" :
		($ENV{TERM} eq 'iris-ansi') ? "\eP1.y$title\e\\" :
		($ENV{TERM} eq 'sun-cmd')   ? "\e]l$title\e\\"   : undef ;
	print { $$self{OUT} } $string if $string;
}

1;

__END__

=head1 NAME

Term::ReadLine::Bender::Base - atomic routines

=head1 DESCRIPTION

This module contains some atomic operations used by all
Term::ReadLine::Bender objects. It is intended as a base class.

At the very least, to child class needs to define a C<default()> function
to handle key bindings and a C<draw()> function
which in turn calls C<print()>. Also the attributes C<IN> and C<OUT>
should contain valid filehandles.

=head1 METHODS

=head2 ANSI stuff

=over 4

=item C<cursor_at($x, $y)>

Positions the cursor on screen, dimensions are 1-based.

=item C<clear_screen()>

Clear screen.

=item C<title($string)>

Set terminal title to C<$string>. When using for example xterm(1)
this is the window name.

=item C<print_length($string)>

Returns the printable length of $string, not counting (some) ansi sequences.

=back

=head2 Private api

Methods for use in overload classes.
I<Avoid using these methods from the application.>

=over 4

=item C<bell()>

Notify the user of an error or limit.

=item C<loop()>

Low level function used by L<readline>. 
Calls C<draw()> and C<do_key()>.

=item C<beat()>

Method called by intervals while waiting for input, to be overloaded.

=item C<read_key()>

Returns one key read from input (this is the named key, not the char when mapped).

=item C<do_key($key)>

Execute a key, calls subroutine for a key binding or the default binding.
If C<$key> is undefined C<read_key()> is called first.

=item C<press($string)>

Do chars in C<$string> like they were typed on the keyboard. Used for testing puposes
and to make macros possible.

If you give more then one argument, these are considered individual characters, use this to
press named keys.

=item C<unread_key($string)>

Unshifts characters on the read buffer, arguments the same as C<press()>.

=item C<key_name($chr)>

Returns a name for a character or character sequence.

=item C<key_binding($key, $mode)>

Returns the keybinding for C<$key> in C<$mode>, mode defaults to the current one.

=item C<bindchr($chr, $key)>

Bind a key name to a character, or a character sequence. All bindings of this kind
are global (you're using only one keyboard, right ?).

=item C<recalc_chr_map()>

Recalculates the chr map, you need to call this after deleting from C<%chr_names>.

=item C<print($lines, $pos)>

Low level function used by L<draw>. Both arguments need to be array references.

=back

=head1 BUGS

Undefined behaviour when the buffer has more lines then the terminal.

Please mail the author if you find any other bugs.

=head1 AUTHOR

Jaap Karssenberg || Pardus [Larus] E<lt>pardus@cpan.orgE<gt>

Copyright (c) 2004 Jaap G Karssenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Term::ReadLine::Bender>

=cut

